rustls/src/msgs/deframer.rs

```diff
@@ -271,8 +278,16 @@ impl MessageDeframer {
                 // Write it into the buffer and create the metadata.
 
                 let expected_len = payload_size(payload)?;
-                let dst = &mut self.buf[..payload.len()];
+                let buf = if quic {
+                    self.buffer.unfilled()
+                } else {
+                    self.buffer.filled_mut()
+                };
+                let dst = &mut buf[..payload.len()];
                 dst.copy_from_slice(payload);
+                if quic {
+                    self.buffer.advance(payload.len());
+                }
```

Comment:
I'd suggest that all of this should probably become a single method on `DeframerBuffer`.